mcproto/types/vec3.py

```diff
@@ -0,0 +1,187 @@
+from __future__ import annotations
+
+import math
+from typing import final
+
+from attrs import Attribute, define, field, validators
+from typing_extensions import override
+
+from mcproto.buffer import Buffer
+from mcproto.protocol import StructFormat
+from mcproto.types.abc import MCType
+
+
+@define
+class Vec3(MCType):
+    """Represents a 3D vector.
+
+    :param x: The x component.
+    :param y: The y component.
+    :param z: The z component.
+    """
+
+    @staticmethod
+    def finite_validator(instance: Vec3, attribute: Attribute[float], value: float) -> None:
+        """Validate that the quaternion components are finite."""
+        if not math.isfinite(value):
+            raise ValueError(f"Quaternion components must be finite, got {value!r}")
+
+    x: float = field(validator=[validators.instance_of((float, int)), finite_validator.__get__(object)])
+    y: float = field(validator=[validators.instance_of((float, int)), finite_validator.__get__(object)])
+    z: float = field(validator=[validators.instance_of((float, int)), finite_validator.__get__(object)])
```

Comment:
What's the point of that `finite_validator.__get__(object)`? From my quick testing, it seems that you can just use `finite_validator` on it's own and it will work just fine.

Same remark for the `Quaternion` class which also does this, and any other occurences of this which I might've missed.

![image](https://github.com/user-attachments/assets/5cdc9ca4-1f4f-4d21-822a-72d44ad311c6)

Reply:
Because it is a static method, not a regular function, in python 3.8 the `__get__` part is needed to get a function because we are getting it from inside the class at creation time.

Reply:
Yeah I figured it's to get the function, I didn't realize it didn't work on 3.8 though, okay then but do add a comment explaining this in all occurrences of doing this.

Alternatively, I wouldn't even mind moving the functions outside of the classes.